libnd4j/blas/NDArray.hpp

```diff
@@ -1719,20 +2262,133 @@ NDArray NDArray::asT() const{
+        stop = nInputoffsets[e + 1];
+        if (dataType() == DataType::UTF8) {
+            dataLength +=  (dtype == DataType::UTF16) ? unicode::offsetUtf8StringInUtf16(data + start, stop)
+                        : unicode::offsetUtf8StringInUtf32(data + start, stop);
+        }
+        else if (dataType() == DataType::UTF16) {
+            dataLength += (dtype == DataType::UTF32) ? unicode::offsetUtf16StringInUtf32(data + start, (stop / sizeof(char16_t)) )
+                        : unicode::offsetUtf16StringInUtf8(data + start, (stop / sizeof(char16_t)));
+        }
+        else {
+            dataLength += (dtype == DataType::UTF16) ? unicode::offsetUtf32StringInUtf16(data + start, (stop / sizeof(char32_t)))
+                        : unicode::offsetUtf32StringInUtf8(data + start, (stop / sizeof(char32_t)));
+        }
+    }
+    offsets[lengthOf()] = dataLength;
+
+    std::shared_ptr<DataBuffer> pBuffer = std::make_shared<DataBuffer>(offsetsLength + dataLength, dtype, getContext()->getWorkspace(), true);
+
+    NDArray res(pBuffer, ShapeDescriptor(dtype, ordering(), getShapeAsVector()), getContext());
+    res.setAttached(getContext()->getWorkspace() != nullptr);
+    
+    if (res.lengthOf() != lengthOf())
+        throw std::invalid_argument("NDArray::asS: Number of strings should match length of array");
+
+    preparePrimaryUse({ &res }, { this });
+
+    memcpy(res.bufferAsT<int8_t>(), offsets.data(), offsets.size() * sizeof(Nd4jLong));
+
+    auto outData = res.bufferAsT<int8_t>() + offsetsLength;
+    const auto inData = bufferAsT<int8_t>() + offsetsLength;
+
+    for (int e = 0; e < lengthOf(); e++) {
+        auto cdata = outData + offsets[e];
+        start = nInputoffsets[e];
+        stop = nInputoffsets[e + 1];
+        auto idata = inData + start;
+        if (dtype == DataType::UTF16) {
+            if (dataType() == DataType::UTF8) {
+                unicode::utf8to16(idata, outData, stop);
+            }
+            else {
+                unicode::utf32to16(idata, outData, (stop / sizeof(char32_t)));
+            }
+        }
+        else if (dtype == DataType::UTF32) {
+            if (dataType() == DataType::UTF8) {
+                unicode::utf8to32(idata, cdata, stop);
+            }
+            else {
+                unicode::utf16to32(idata, outData, (stop / sizeof(char16_t)));
+            }
+        }
+        else {
+            if (dataType() == DataType::UTF16) {
+                unicode::utf16to8(idata, outData, (stop / sizeof(char16_t)));
+            }
+            else {
+                unicode::utf32to8(idata, outData, (stop / sizeof(char32_t)));
+            }
+        }
+    }
+
+    registerPrimaryUse({ &res }, { this });
+
+    return res;
+}
+BUILD_SINGLE_TEMPLATE(template ND4J_EXPORT NDArray NDArray::asS, () const, LIBND4J_STRINGTYPES);
+
 ////////////////////////////////////////////////////////////////////////
 NDArray NDArray::asT(DataType dtype) const {
-    if (isS())
```

Comment:
We still need this kind of validation, just to make sure nobody tries to cast float to utf. Bring it back please within your if/else branches

Reply:
fixed